pkg/apis/rollouts/validation/validation.go

```diff
@@ -0,0 +1,268 @@
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	//"github.com/argoproj/argo-rollouts/utils/defaults"
+)
+
+const (
+	// Validate Spec constants
+
+	// InvalidSpecReason indicates that the spec is invalid
+	InvalidSpecReason = "InvalidSpec"
+	// MissingFieldMessage the message to indicate rollout is missing a field
+	MissingFieldMessage = "Rollout has missing field '%s'"
+	// RolloutSelectAllMessage the message to indicate that the rollout has an empty selector
+	RolloutSelectAllMessage = "This rollout is selecting all pods. A non-empty selector is required."
+	// InvalidSetWeightMessage indicates the setweight value needs to be between 0 and 100
+	InvalidSetWeightMessage = "SetWeight needs to be between 0 and 100"
+	// InvalidDurationMessage indicates the Duration value needs to be greater than 0
+	InvalidDurationMessage = "Duration needs to be greater than 0"
+	// InvalidMaxSurgeMaxUnavailable indicates both maxSurge and MaxUnavailable can not be set to zero
+	InvalidMaxSurgeMaxUnavailable = "MaxSurge and MaxUnavailable both can not be zero"
+	// InvalidStepMessage indicates that a step must have either setWeight or pause set
+	InvalidStepMessage = "Step must have one of the following set: experiment, setWeight, or pause"
+	// ScaleDownDelayLongerThanDeadlineMessage indicates the ScaleDownDelaySeconds is longer than ProgressDeadlineSeconds
+	ScaleDownDelayLongerThanDeadlineMessage = "ScaleDownDelaySeconds cannot be longer than ProgressDeadlineSeconds"
+	// RolloutMinReadyLongerThanDeadlineMessage indicates the MinReadySeconds is longer than ProgressDeadlineSeconds
+	RolloutMinReadyLongerThanDeadlineMessage = "MinReadySeconds cannot be longer than ProgressDeadlineSeconds"
+	// InvalidStrategyMessage indiciates that multiple strategies can not be listed
+	InvalidStrategyMessage = "Multiple Strategies can not be listed"
+	// DuplicatedServicesMessage the message to indicate that the rollout uses the same service for the active and preview services
+	DuplicatedServicesMessage = "This rollout uses the same service for the active and preview services, but two different services are required."
+	// ScaleDownLimitLargerThanRevisionLimit the message to indicate that the rollout's revision history limit can not be smaller than the rollout's scale down limit
+	ScaleDownLimitLargerThanRevisionLimit = "This rollout's revision history limit can not be smaller than the rollout's scale down limit"
+	// AvailableReason the reason to indicate that the rollout is serving traffic from the active service
+	AvailableReason = "AvailableReason"
+	// NotAvailableMessage the message to indicate that the Rollout does not have min availability
+	NotAvailableMessage = "Rollout does not have minimum availability"
+	// AvailableMessage the message to indicate that the Rollout does have min availability
+	AvailableMessage = "Rollout has minimum availability"
+)
+
+// called in strategy.go -> create.go
```

Comment:
Can you elaborate on this comment more?

Reply:
These are just my personal notes about how ValidateDeployment is used in the Kubernetes library.